Source/WebCore/Modules/url-pattern/URLPatternConstructorStringParser.cpp

```diff
@@ -0,0 +1,370 @@
+        // Look for password prefix
+        if (isNonSpecialPatternChar(m_tokenIndex, ':'))
+            changeState(URLPatternConstructorStringParserState::Password, 1);
+        // Look for identity terminator
+        else if (isNonSpecialPatternChar(m_tokenIndex, '@'))
+            changeState(URLPatternConstructorStringParserState::Hostname, 1);
+        break;
+    case URLPatternConstructorStringParserState::Password:
+        // Look for identity terminator
+        if (isNonSpecialPatternChar(m_tokenIndex, '@'))
+            changeState(URLPatternConstructorStringParserState::Hostname, 1);
+        break;
+    case URLPatternConstructorStringParserState::Hostname:
+        // Look for an IPv6 open
+        if (isNonSpecialPatternChar(m_tokenIndex, '['))
+            m_hostnameIPv6BracketDepth += 1;
+        // Look for an IPv6 close
+        else if (isNonSpecialPatternChar(m_tokenIndex, ']'))
+            m_hostnameIPv6BracketDepth -= 1;
+        // Look for port prefix
+        else if (isNonSpecialPatternChar(m_tokenIndex, ':'))
+            changeState(URLPatternConstructorStringParserState::Port, 1);
+        // Look for pathname start
+        else if (isNonSpecialPatternChar(m_tokenIndex, '/'))
+            changeState(URLPatternConstructorStringParserState::Pathname, 0);
+        // Look for search prefix
+        else if (isSearchPrefix())
+            changeState(URLPatternConstructorStringParserState::Search, 1);
+        // Look for hash prefix
+        else if (isNonSpecialPatternChar(m_tokenIndex, '#'))
+            changeState(URLPatternConstructorStringParserState::Hash, 1);
+        break;
+    case URLPatternConstructorStringParserState::Port:
+        // Look for pathname start
+        if (isNonSpecialPatternChar(m_tokenIndex, '/'))
+            changeState(URLPatternConstructorStringParserState::Pathname, 0);
+        else if (isSearchPrefix())
+            changeState(URLPatternConstructorStringParserState::Search, 1);
+        // Look for hash prefix
+        else if (isNonSpecialPatternChar(m_tokenIndex, '#'))
+            changeState(URLPatternConstructorStringParserState::Hash, 1);
+        break;
+    case URLPatternConstructorStringParserState::Pathname:
+        if (isSearchPrefix())
+            changeState(URLPatternConstructorStringParserState::Search, 1);
+        // Look for hash prefix
+        else if (isNonSpecialPatternChar(m_tokenIndex, '#'))
+            changeState(URLPatternConstructorStringParserState::Hash, 1);
+        break;
+    case URLPatternConstructorStringParserState::Search:
+        // Look for hash prefix
+        if (isNonSpecialPatternChar(m_tokenIndex, '#'))
+            changeState(URLPatternConstructorStringParserState::Hash, 1);
+        break;
+    case URLPatternConstructorStringParserState::Hash:
+        // Do nothing
+        break;
+    case URLPatternConstructorStringParserState::Done:
+        // This step is never reached
+        ASSERT_NOT_REACHED();
```

Comment:
Comments like these are not needed though, since the code is quite self-explanatory.